keps/sig-network/2732-np-version/README.md

```diff
@@ -0,0 +1,803 @@
+how kube-proxy implements Service rules). In this case it is not clear
+who would be responsible for updating `NetworkPolicyStatus`, and it may
+not be easy for any part of the system to know with certainty that a
+given NetworkPolicy was fully in effect across the cluster.
+
+In other cases, a plugin may implement NetworkPolicy in terms of some
+other lower-level policy engine (implemented by a cloud, perhaps), and
+may work by simply translating NetworkPolicy rules into rules for this
+other engine. As in the distributed case, it may not be possible for
+such a NetworkPolicy implementation to know for sure when a policy was
+fully in effect.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+Add a new field to `networkingv1.NetworkPolicySpec`:
+
+```
+type NetworkPolicySpec struct {
+	...
+
+	// minVersion indicates the Kubernetes release version corresponding to
+	// the NetworkPolicy functionality required by this policy. If this is
+	// explicitly specified by the user, the apiserver will require that the policy
+	// conforms to the specified version. If it is not specified, the apiserver
+	// will fill in the correct minVersion based on the features used by the policy.
+	// +optional
+	MinVersion NetworkPolicyVersion `json:"minVersion,omitempty" protobuf:"bytes,5,name=minVersion"`
+
+	...
+}
+
+type NetworkPolicyVersion string
+
+const (
+	// Kubernetes 1.3 introduced the NetworkPolicy v1beta1 API, supporting
+	// ingress-only policies containing podSelectors and namespaceSelectors,
+	// and TCP and UDP ports.
+	NetworkPolicyVersion1_3 NetworkPolicyVersion = "1.3"
+
+	// Kubernetes 1.8 brought the NetworkPolicy API to v1, adding egress policies
+	// and ipBlocks.
+	NetworkPolicyVersion1_8 NetworkPolicyVersion = "1.8"
+
+	// Kubernetes 1.9 added alpha IPv6 support. Plugins implementing this version are
+	// assumed to recognize IPv6 addresses, even if they don't implement them.
+	NetworkPolicyVersion1_9 NetworkPolicyVersion = "1.9"
+
+	// Kubernetes 1.11 added support for specifying podSelector and namespaceSelector
+	// together in a peer.
+	NetworkPolicyVersion1_11 NetworkPolicyVersion = "1.11"
+
+	// Kubernetes 1.12 added alpha SCTP support. Plugins implementing this version are
+	// assumed to recognize `protocol: SCTP`, even if they don't implement it.
+	NetworkPolicyVersion1_12 NetworkPolicyVersion = "1.12"
+)
+
+```
+
+And add a new `Status NetworkPolicyStatus` field to `NetworkPolicy`, defined as:
+
+```
+// NetworkPolicyStatus contains information about the processing of a NetworkPolicy
+type NetworkPolicyStatus struct {
+	// conditions associated with the policy. These may include a "Supported"
+	// condition indicating whether the features used by the policy are supported,
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []NetworkPolicyStatusCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,3,rep,name=conditions"`
```

Comment:
FTR I used Conditions in the combined Status+Versioning KEP, but it's not necessarily clear that Conditions are right for just Versioning.